### PR TITLE
Implements Persian calendar

### PR DESCRIPTION
**Pull request by jason-p-pickering**

### CODE REVIEW

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/period/QuarterlyPeriodType.java [hunk 1]
```diff
@@ -180,6 +180,11 @@ public Period getPreviousPeriod( Period period, org.hisp.dhis.calendar.Calendar
     @Override
     public String getIsoDate( DateTimeUnit dateTimeUnit, org.hisp.dhis.calendar.Calendar calendar  )
     {
+
+        if (calendar.name() == "persian" && dateTimeUnit.isIso8601()) {
```
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/period/QuarterlyPeriodType.java]: This looks a bit hacky. Can we make it more dynamic?

Also please follow Java code style.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/calendar/impl/PersianCalendar.java [hunk 2]
```diff
@@ -0,0 +1,683 @@
...
+import org.joda.time.DateTimeZone;
+import org.joda.time.Days;
+import org.joda.time.chrono.ISOChronology;
+import org.springframework.stereotype.Component;
+
+import org.hisp.dhis.calendar.exception.InvalidCalendarParamatersException;
+/**
+ * @author Hans Jacobson <jacobson.hans@gmail.com>
+ */
+
```
Comment (mortenoh) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/calendar/impl/PersianCalendar.java]: Extra newline

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/calendar/impl/PersianCalendar.java [hunk 3]
```diff
@@ -0,0 +1,683 @@
...
+@Component
+public class PersianCalendar
+    extends AbstractCalendar
+{
+
```
Comment (mortenoh) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/calendar/impl/PersianCalendar.java]: Extra newline

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/calendar/impl/PersianCalendar.java [hunk 4]
```diff
@@ -0,0 +1,683 @@
...
+    //
+    //------------------------------------------------------------------------------------------------------------
+
+    /**
+     * Map that gives an array of month lengths based on Persian year lookup.
+     * Index 1 - 12 is used for months, index 0 is used to give year total (lazy calculated).
+     */
+    private static final Map<Integer, int[]> CONVERSION_MAP = new HashMap<>();
+
+    static
```
Comment (mortenoh) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/calendar/impl/PersianCalendar.java]: Maybe move this to end of file, normally static defs are on top.. but this one is big

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/calendar/impl/PersianCalendar.java [hunk 5]
```diff
@@ -0,0 +1,683 @@
...
+    @Override
+    public String name()
+    {
+        return "persian";
+    }
+
+    @Override
+    public DateTimeUnit toIso( DateTimeUnit dateTimeUnit )
+    {
+
```
Comment (mortenoh) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/calendar/impl/PersianCalendar.java]: Extra newline

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/calendar/impl/PersianCalendar.java [hunk 6]
```diff
@@ -0,0 +1,683 @@
...
+        {
+            totalDays += getDaysFromMap( dateTimeUnit.getYear(), month );
+        }
+
+        totalDays += dateTimeUnit.getDay() - START_PERSIAN.getDay();
+
+        dateTime = dateTime.plusDays( totalDays );
+
+        return new DateTimeUnit( DateTimeUnit.fromJodaDateTime( dateTime ), true );
+
```
Comment (mortenoh) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/calendar/impl/PersianCalendar.java]: Extra newline

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/calendar/impl/PersianCalendar.java [hunk 7]
```diff
@@ -0,0 +1,683 @@
...
+            throw new InvalidCalendarParamatersException(
+                "Illegal ISO year, must be between " + START_ISO.getYear() + " and " + STOP_ISO.getYear() +
+                    ", was given " + dateTimeUnit.getYear() );
+        }
+
+        DateTime start = START_ISO.toJodaDateTime();
+        DateTime end = dateTimeUnit.toJodaDateTime();
+
+        return plusDays( START_PERSIAN, Days.daysBetween( start, end ).getDays() );
+
```
Comment (mortenoh) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/calendar/impl/PersianCalendar.java]: Extra newline

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/calendar/impl/PersianCalendar.java [hunk 8]
```diff
@@ -0,0 +1,683 @@
...
+            case ISO8601_MONTH:
+                return toMonthIsoInterval( dateTimeUnit, offset, length );
+            case ISO8601_WEEK:
+                return toWeekIsoInterval( dateTimeUnit, offset, length );
+            case ISO8601_DAY:
+                return toDayIsoInterval( dateTimeUnit, offset, length );
+            default:
+                return null;
+        }
+
```
Comment (mortenoh) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/calendar/impl/PersianCalendar.java]: Extra newline

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/calendar/impl/PersianCalendar.java [hunk 9]
```diff
@@ -0,0 +1,683 @@
...
+        DateTime dateTime = toIso( dateTimeUnit )
+            .toJodaDateTime( ISOChronology.getInstance( DateTimeZone.getDefault() ) );
+        return dateTime.getDayOfWeek();
+    }
+
+    @Override
+    public int weekday( DateTimeUnit dateTimeUnit )
+    {
+
```
Comment (mortenoh) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/calendar/impl/PersianCalendar.java]: Extra newline

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/calendar/impl/PersianCalendar.java [hunk 10]
```diff
@@ -0,0 +1,683 @@
...
+            {
+                CONVERSION_MAP.get( year )[0] += CONVERSION_MAP.get( year )[j];
+            }
+        }
+
+        return CONVERSION_MAP.get( year )[0];
+    }
+
+    @Override
+    public DateTimeUnit minusYears( DateTimeUnit dateTimeUnit, int years )
```
Comment (mortenoh) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/calendar/impl/PersianCalendar.java]: Make sure that public methods come first in the file, not a mix of public/private..

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/calendar/impl/PersianCalendar.java [hunk 11]
```diff
@@ -0,0 +1,683 @@
...
+        int dm = getDaysFromMap( result.getYear(), result.getMonth() );
+
+        if ( result.getDay() > dm )
+        {
+            result.setDay( dm );
+        }
+
+        result.setDayOfWeek( weekday( result ) );
+    }
+
```
Comment (mortenoh) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/calendar/impl/PersianCalendar.java]: Extra newlines

File: dhis-2/dhis-api/src/test/java/org/hisp/dhis/calendar/impl/PersianCalendarTest.java [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+
+import org.junit.Before;
+import org.junit.Test;
+
+/**
+ * @author Hans Jacobson <jacobson.hans@gmail.com>
+ */
+public class PersianCalendarTest
+{
+
```
Comment (mortenoh) [inline dhis-2/dhis-api/src/test/java/org/hisp/dhis/calendar/impl/PersianCalendarTest.java]: Extra newline

**Pull request merged**